x-pack/test/detection_engine_api_integration/utils/create_prebuilt_rule_so.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Client } from '@elastic/elasticsearch';
+import { ToolingLog } from '@kbn/tooling-log';
+import {
+  getPrebuiltRuleMock,
+  getPrebuiltRuleWithExceptionsMock,
+} from '@kbn/security-solution-plugin/common/detection_engine/prebuilt_rules/model/prebuilt_rule.mock';
+import { countDownTest } from './count_down_test';
+
+export const ELASTIC_SECURITY_RULE_ID = '9a1a2dae-0b5f-4c3d-8305-a268d404c306';
+
+export const SAMPLE_PREBUILT_RULES = [
+  {
+    'security-rule': {
+      ...getPrebuiltRuleWithExceptionsMock(),
+      rule_id: ELASTIC_SECURITY_RULE_ID,
+    },
+    type: 'security-rule',
+    references: [],
+    coreMigrationVersion: '8.6.0',
+    updated_at: '2022-11-01T12:56:39.717Z',
+    created_at: '2022-11-01T12:56:39.717Z',
+  },
+  {
+    'security-rule': {
+      ...getPrebuiltRuleMock(),
+      rule_id: '000047bb-b27a-47ec-8b62-ef1a5d2c9e19',
+    },
+    type: 'security-rule',
+    references: [],
+    coreMigrationVersion: '8.6.0',
+    updated_at: '2022-11-01T12:56:39.717Z',
+    created_at: '2022-11-01T12:56:39.717Z',
+  },
+  {
+    'security-rule': {
+      ...getPrebuiltRuleMock(),
+      rule_id: '00140285-b827-4aee-aa09-8113f58a08f3',
+    },
+    type: 'security-rule',
+    references: [],
+    coreMigrationVersion: '8.6.0',
+    updated_at: '2022-11-01T12:56:39.717Z',
+    created_at: '2022-11-01T12:56:39.717Z',
+  },
+];
+
+/**
+ * Creates saved objects with prebuilt rules for further installation
+ *
+ * @param es Elasticsearch client
+ * @param log Log client
+ */
+export const createPrebuiltRuleSOs = async (es: Client, log: ToolingLog): Promise<void> => {
+  await countDownTest(
```

Comment:
> Why do we need a retry logic for creating the rule asset saved objects?

I'm not sure about that, to be honest. Probably there were some race conditions causing tests to fail, but I see many (if not all) idempotent calls are wrapped either in `countDownTest` or `countDownES`. So I decided to follow suit. There's a JSDoc description for these helpers:

> Does a plain countdown and checks against a boolean to determine if to wait and try again. This is useful for over the wire things that can cause issues such as conflict or timeouts for testing resiliency.

That said, I'm okay with removing those retries and adding them only once we see any problems with the stability of the tests.

> Nit: Probably an offtopic. This function has retry count == 50, wondering if that's not too much for a default value.

IMO too much, yes. We could set it to something like 10. But, on the other hand, as our previous experience with timeouts in tests shows, a too-small value could lead to random failures in CI and locally.

Reply:
> I'm okay with removing those retries and adding them only once we see any problems with the stability of the tests.

I'd remove the retry until we have a need for it and can make an informed decision on using it.

> IMO too much, yes. We could set it to something like 10. But, on the other hand, as our previous experience with timeouts in tests shows, a too-small value could lead to random failures in CI and locally.

Let's keep it but we might want to revisit this value in the future.

Overall I think my idea is: having retry logic all over the place in the integration tests can bring more harm than value because:

- It can hide issues and bugs, e.g. it can hide issues with race conditions and OCC caused by our BE logic. Our app doesn't retry everything, so we might end up with green tests but a broken app. As an example, I remember we had a race condition between a rule executor and user actions when both were trying to update the rule status, and user actions or rule executors were sometimes failing with an OCC error. Having retry logic "by default" will hide such bugs.
- Having a large number of retries can make broken tests run significantly longer.

I would probably use retry logic in tests only in the following cases:

- To implement long-polling checks ("wait until this alert is indexed", "wait until the rule status is failed", etc)
- To handle specific issues with stability (e.g. let's say there's a heavy operation like installing a fleet integration and it fails sometimes and we know why it fails and it is related to the test environment and not to bugs in the app)

Reply:
Okay, I removed the retries, and the build has succeeded without any failures. So I think we're good here 👍
Let's keep an eye on future CI builds. If we see no extra flakiness, we could also revisit and remove other retries.